services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/converter/attribute/ChainAttributeConverter.java

```diff
@@ -0,0 +1,265 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.enhanced.dynamodb.converter.attribute;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import software.amazon.awssdk.annotations.Immutable;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.annotations.ThreadSafe;
+import software.amazon.awssdk.enhanced.dynamodb.converter.attribute.UnionAttributeConverter.Visitor;
+import software.amazon.awssdk.enhanced.dynamodb.model.AttributeConverterAware;
+import software.amazon.awssdk.enhanced.dynamodb.model.ItemAttributeValue;
+import software.amazon.awssdk.enhanced.dynamodb.model.TypeToken;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * A chain of converters, invoking the underlying converters based on the precedence defined in the
+ * {@link AttributeConverterAware} documentation.
+ *
+ * <p>
+ * Given an input, this will identify a converter that can convert the specific Java type and invoke it. If a converter cannot
+ * be found, it will invoke a "parent" converter, which would be expected to be able to convert the value (or throw an exception).
+ */
+@SdkPublicApi
+@ThreadSafe
+@Immutable
+public final class ChainAttributeConverter<T> implements SubtypeAttributeConverter<T> {
+    private static final Logger log = Logger.loggerFor(ChainAttributeConverter.class);
+
+    private final TypeToken<T> type;
+
+    private final List<SubtypeAttributeConverter<? extends T>> subtypeConverters = new ArrayList<>();
+
+    private final ConcurrentHashMap<Class<? extends T>, UnionAttributeConverter<? extends T>> converterCache =
+            new ConcurrentHashMap<>();
+
+    /**
+     * The "default converter" to invoke if no converters can be found in this chain supporting a specific type.
+     */
+    private final SubtypeAttributeConverter<? super T> parent;
+
+    private ChainAttributeConverter(Builder<T> builder) {
+        this.type = builder.type;
+        this.parent = builder.parent;
+
+        // Subtype converters are used in the REVERSE order of how they were added to the builder.
+        for (int i = builder.subtypeConverters.size() - 1; i >= 0; i--) {
+            this.subtypeConverters.add(builder.subtypeConverters.get(i));
+        }
+
+        // Converters are used in the REVERSE order of how they were added to the builder.
```

Comment:
I can clarify in the comments. The implementation iterates the list in order, so we need to reverse it to match the contract defined in `AttributeConverterAware`:

> Converters added last always have a higher precedence than those added first. For example, the last {@code AttributeConverter<String>)} converter added to a {@link RequestItem.Builder} will have higher precedence than any other {@code AttributeConverter<String>} converters that were added before it.

It doesn't affect the VAST majority of customers, but the reasoning for that is because the code closest to the call site should have the ability to override the behavior of earlier code, because it's more aware of the context around the request. E.g. If they have global logic that pre-configures the client, they could want to override that behavior later.